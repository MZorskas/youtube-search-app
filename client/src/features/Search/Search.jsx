import React, { useState, useEffect } from 'react';
import './Search.scss';
import { inputRegex } from 'utils/regex';
import IconButton from 'components/IconButton';
import ErrorMessage from 'components/ErrorMessage';
import Input from 'components/Input';
import { SearchIcon } from 'assets/icons';
import PropTypes from 'prop-types';
import { LOGGER_ACTIONS } from 'loggerActions';

const Search = ({ handleSearch, logUserAction }) => {
  const [inputValue, setInputValue] = useState();
  const [validationError, setValidationError] = useState('');

  useEffect(() => {
    setValidationError('');
  }, [inputValue]);

  const handleSubmit = (event) => {
    event.preventDefault();

    setValidationError('');

    const queryDetails = {
      keywords: inputValue,
    };

    logUserAction(LOGGER_ACTIONS.SEARCH, queryDetails);

    if (!inputValue) return setValidationError('Input should not be empty');

    if (inputValue.length > 20)
      return setValidationError(
        'Input should not be longer than 20 characters'
      );

    if (!inputRegex.test(inputValue))
      return setValidationError('Input should only contain letters');

    handleSearch(inputValue);
  };

  const handleChange = (event) => {
    setInputValue(event.target.value);
  };

  return (
    <section className="search" data-testid="search">
      <form
        className="search__form"
        aria-label="search"
        onSubmit={handleSubmit}
        data-testId="search-form"
      >
        <Input
          className="search__input"
          onChange={handleChange}
          placeholder="Videos"
          validationError={!!validationError}
          id="search-input"
          label="Search"
          autoFocus
        />
        <IconButton
          className="search__button"
          icon={<SearchIcon className="search__icon" />}
          type="submit"
        />
      </form>
      {validationError && (
        <ErrorMessage
          className="search__error-message"
          message={validationError}
        />
      )}
    </section>
  );
};

Search.propTypes = {
  handleSearch: PropTypes.func.isRequired,
  logUserAction: PropTypes.func.isRequired,
};

export default Search;
