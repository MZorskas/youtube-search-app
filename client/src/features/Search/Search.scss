@import 'src/styles/mixins/index';
@import 'src/styles/base/base';

.search {
  min-height: 9.4rem;

  &__form {
    display: flex;
  }
  &__input {
    $_input-height: 6rem;

    min-height: $_input-height;
    padding: 0 1.6rem;
    width: 100%;
    transition: 0.5s;

    &::placeholder {
      @include text-style-h4;

      color: $color-font-secondary;
      font-weight: $font-weight-regular;
      line-height: $_input-height;
    }
  }

  &__button {
    padding: 0 1.3rem;
  }

  &__icon {
    $_icon-size: 3.4rem;

    width: $_icon-size;
    height: $_icon-size;
  }

  &__error-message {
    margin: 0.5rem 0;
  }

  @include if-phone () {
    text-align: center;

    &__input {
      $_input-height: 5rem;

      min-height: $_input-height;

      &::placeholder {
        line-height: $_input-height;
      }
    }

    &__button {
      padding: 0 1.2rem;
    }

    &__icon {
      width: 2.6rem;
      height: 2.6rem;
    }
  }
}
