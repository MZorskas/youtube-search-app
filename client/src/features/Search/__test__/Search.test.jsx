import React from 'react';
import Search from '../Search';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const mockHandleSearch = jest.fn();
const mockLogData = jest.fn();

const inputSetup = () => {
  const utils = render(
    <Search handleSearch={mockHandleSearch} logUserAction={mockLogData} />
  );
  const input = screen.queryByPlaceholderText('Videos');
  const form = screen.queryByTestId('search-form');
  return {
    input,
    form,
    ...utils,
  };
};

describe('Search component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');

    expect(screen.queryByTestId('search')).not.toBeInTheDocument();
    render(<Search />);
    expect(screen.queryByTestId('search')).toBeInTheDocument();

    expect(spy).toHaveBeenCalled();
  });

  it('renders without propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');

    expect(screen.queryByTestId('search')).not.toBeInTheDocument();
    render(<Search handleSearch={mockHandleSearch} />);
    expect(screen.queryByTestId('search')).toBeInTheDocument();

    expect(spy).not.toHaveBeenCalled();
  });

  // it('renders and focuses inputField', () => {
  //   expect(screen.queryByTestId('search')).not.toBeInTheDocument();
  //   render(<Search handleSearch={mockHandleSearch} />);
  //   expect(screen.queryByTestId('search')).toBeInTheDocument();
  //   expect(screen.queryByPlaceholderText('Gifs')).toHaveFocus();
  // });

  it('fires handleSubmit function on submit', () => {
    const { input } = inputSetup();
    expect(screen.queryByTestId('search-form')).toBeInTheDocument();
    expect(input).toBeInTheDocument();

    expect(input.value).toBe('');
    expect(mockHandleSearch).toHaveBeenCalledTimes(0);
    fireEvent.submit(input);
    expect(mockHandleSearch).toHaveBeenCalledTimes(0);

    fireEvent.change(input, { target: { value: 'gifs' } });
    expect(input.value).toBe('gifs');
    fireEvent.submit(input);
    expect(mockHandleSearch).toHaveBeenCalledTimes(1);
  });

  it('validates input value on submit', () => {
    const { input, form } = inputSetup();
    expect(form).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(mockHandleSearch).toHaveBeenCalledTimes(0);

    expect(input.value).toBe('');
    fireEvent.submit(input);
    expect(mockHandleSearch).toHaveBeenCalledTimes(0);
    expect(screen.queryByTestId('error-message')).toHaveTextContent(
      'Input should not be empty'
    );

    fireEvent.change(input, { target: { value: '#' } });
    fireEvent.submit(input);
    expect(mockHandleSearch).toHaveBeenCalledTimes(0);
    expect(screen.queryByTestId('error-message')).toHaveTextContent(
      'Input should only contain letters'
    );

    fireEvent.change(input, { target: { value: '1' } });
    fireEvent.submit(input);
    expect(mockHandleSearch).toHaveBeenCalledTimes(0);
    expect(screen.queryByTestId('error-message')).toHaveTextContent(
      'Input should only contain letters'
    );

    fireEvent.change(input, { target: { value: 'gifs' } });
    fireEvent.submit(input);
    expect(mockHandleSearch).toHaveBeenCalledTimes(1);
    expect(screen.queryByTestId('error-message')).not.toBeInTheDocument();
  });

  it('validates clears errorMessage onChange', () => {
    const { input, form } = inputSetup();
    expect(form).toBeInTheDocument();
    expect(input).toBeInTheDocument();

    expect(input.value).toBe('');
    fireEvent.submit(input);
    expect(screen.queryByTestId('error-message')).toHaveTextContent(
      'Input should not be empty'
    );
    fireEvent.change(input, { target: { value: '#' } });
    expect(screen.queryByTestId('error-message')).not.toBeInTheDocument();
    fireEvent.submit(input);
    expect(screen.queryByTestId('error-message')).toHaveTextContent(
      'Input should only contain letters'
    );
    fireEvent.change(input, { target: { value: '#1' } });
    expect(screen.queryByTestId('error-message')).not.toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Search handleSearch={mockHandleSearch} logUserAction={mockLogData} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
