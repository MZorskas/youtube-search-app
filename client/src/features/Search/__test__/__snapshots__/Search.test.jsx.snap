// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component matches snapshot 1`] = `
<section
  className="search"
  data-testid="search"
>
  <form
    aria-label="search"
    className="search__form"
    data-testId="search-form"
    onSubmit={[Function]}
  >
    <div
      className="input-wrapper"
      data-testid="input"
    >
      <label
        className="input-wrapper__label"
        htmlFor="search-input"
      >
        Search
      </label>
      <input
        autoFocus={true}
        className="input-wrapper__input search__input"
        id="search-input"
        onChange={[Function]}
        placeholder="Videos"
      />
    </div>
    <button
      aria-label="search"
      className="icon-button search__button"
      data-testid="icon-button"
      type="submit"
    >
      <svg
        className="search__icon"
      >
        search.svg
      </svg>
    </button>
  </form>
  
</section>
`;
